src/application/downloader.py

```
@@ -0,0 +1,218 @@
+        completed_ids = set(checkpoint.get('completed_locations', [])) if checkpoint else set()
+
+        location_semaphore = asyncio.Semaphore(self.max_concurrent_locations)
+        
+        tasks = []
+        for i, location in enumerate(locations[start_index:], start=start_index):
+            if location.id in completed_ids:
+                continue
+            
+            task = asyncio.create_task(
+                self._download_location_with_limit(
+                    location, job, i, len(locations), location_semaphore, completed_ids
```

Ruff (0.12.2)

Trailing comma missing

Add trailing comma

(COM812)